### ISSUE TITLE
Building binutils ac_fn_c_check_decl fails

### ISSUE THREAD

**Issue by lal12** [emscripten-core/emscripten#7138]
I am trying to build `binutils` (cause I want to embed the strip utility inside a node script).
Mostly this works fine, but I noticed an error regarding `libiberty`, since all functionality of `libiberty` is included in emscripten it does not need to build its own, which leads to problems when compiling the bitcode to wasm.

The problem is that the way `ac_fn_c_check_decl` tryes to detect whether a given function exists is with the following code (e.g. for get_optargs)

```
/* end confdefs.h.  */
/* Define psignal to an innocuous variant, in case <limits.h> declares psignal.
   For example, HP-UX 11i <limits.h> declares gettimeofday.  */
#define psignal innocuous_psignal

/* System header to define __stub macros and hopefully few prototypes,
    which can conflict with char psignal (); below.
    Prefer <limits.h> to <assert.h> if __STDC__ is defined, since
    <limits.h> exists even on freestanding compilers.  */

#ifdef __STDC__
# include <limits.h>
#else
# include <assert.h>
#endif

#undef psignal

/* Override any GCC internal prototype to avoid an error.
   Use char because int might match the return type of a GCC
   builtin and then its argument prototype would still apply.  */
#ifdef __cplusplus
extern "C"
#endif
char psignal ();
/* The GNU C library defines this for functions which it implements
    to always fail with ENOSYS.  Some functions are actually named
    something starting with __ and the normal name is an alias.  */
#if defined __stub_psignal || defined __stub___psignal
choke me
#endif

int main (){
  return psignal ();
  return 0;
}
```
So to summarize it, the code try's to redefine psignal without any arguments, and then tries to call it. By successfully compiling and linking it, it knows that psignal does exists, since the C symbols for linking do not include function arguments (unlike C++ ones).

Of course I could easily fix this by modifying the `configure` of `libiberty`, but I think there should be a better way for solving such problem in general, this case does not seem to be that unusual, since it is a standard autotools function.

**Edit**:
I forgot to append the error message, printed by emscripten:
`unexpected number of arguments 0 in call to 'psignal', should be 1`

**Comment by kripken** [comment]
I think that's just a warning message - is the issue that it sees output to stderr?

**Comment by kripken** [comment]
Hmm, `error: unresolved symbol: psignal` looks like a valid error message - we don't have support for signals, so makes sense it would find it isn't present. So isn't the error the right behavior there?

**Comment by lal12** [comment]
It seems to be at least declared, because I get an error during later
compilation about two different interfering declarations of psignal, and
one of thoose comes from a header file inside emscripten
(`emscripten/1.38.12/system/include/libc/signal.h:219:void psignal(int,
const char *);`

Am Di., 18. Sep. 2018 um 01:01 Uhr schrieb Alon Zakai <
notifications@github.com>:

**Comment by kripken** [comment]
What is the specific error you mention that you get later?

**Comment by lal12** [comment]
```
warning: unknown warning option '-Wshadow=local' [-Wunknown-warning-option]
../../binutils-2.31/libiberty/strsignal.c:554:1: error: conflicting types for
      'psignal'
psignal (int signo, char *message)
^
/home/ll/git/wasm-binutils-arm-linux/emsdk/emscripten/1.38.12/system/include/libc/signal.h:219:6: note: 
      previous declaration is here
void psignal(int, const char *);
     ^
1 warning and 1 error generated.
ERROR:root:compiler frontend failed to generate LLVM bitcode, halting
Makefile:1455: recipe for target 'strsignal.o' failed
make[2]: *** [strsignal.o] Error 1
```
So the problem is the different declaration of psignal (parameter 2 is const char*, while libiberty defines char*). But I still think the correct solution would be to make the test works.

Btw I can compile without problems by setting some internal defines by hand (`CFLAGS="-DHAVE_PSIGNAL=1 -DELIDE_CODE -D__GNU_LIBRARY__"`), but I think the error in the `ac_fn_c_check_decl` function is more of a fundamental problem...

**Comment by kripken** [comment]
Emscripten's `signal.h` is from musl, and it looks like musl is right in defining it as const (e.g. http://man7.org/linux/man-pages/man3/psignal.3.html ). So this looks like a bug in libiberty (which should either leave defining it to the system headers, or define it in the standard way). However, I'm not an expert on these headers, so I may be missing something.

**Comment by lal12** [comment]
Well, I don't think it matters whether musl or libiberty is right, instead the underlying problem happening during the configure of libiberty should be the problem to look at, since libiberty assumes that psignal is not defined elsewhere, they should be allowed to define it however they want.

**Comment by kripken** [comment]
Does libiberty both define and implement `psignal` itself? If it does, then I'm not sure how that would work with linking to an existing libc. If it only defines and doesn't implement, it seems like it has to define it in a way consistent with libcs - and that appears to not be done correctly.

It's possible the reason this shows up in emscripten and not elsewhere is we build with a fairly new version of clang, that may have more errors on things like this. In other words, building libiberty natively with the same clang should hit the same error - I think.

**Comment by lal12** [comment]
`libiberty` does declare and define `psignal` (and other `libc` functions), but only if it not detects thoose functions as existing, so it usually should not conflict with a libc if the detection during the configure (which fails here) works correctly. Basically a large part of `libiberty` is a polyfill to make a program work either without a libc or a libc with a reduced feature set.

**Comment by kripken** [comment]
Is the problem that emscripten defines psignal in a header, but doesn't have it in the actual library implementation? If so, we can resolve that by adding stubs for them in `src/library_signal.js` (that throw if they are called, or return a proper error message).

**Comment by Gamadril** [comment]
I have also a problem with psignal. It seems that `emconfigure configure` does not recognize emscriptens's psignal: `checking for psignal... no`. So `emcmake make` compiles own psignal implementation and fails later: 

```
./strsignal.c:554:1: error: conflicting types for 'psignal'

psignal (int signo, char *message)

../emsdk/fastcomp/emscripten/system/include/libc/signal.h:219:6: note: previous declaration is here
void psignal(int, const char *);

1 warning and 1 error generated.
shared:ERROR: compiler frontend failed to generate LLVM bitcode, halting
```
